src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp

```diff
@@ -0,0 +1,319 @@
+    DINLINE unsigned int numNewParticles(FrameType& sourceFrame, int localIdx)
+    {
+        using namespace PMacc::algorithms;
+
+        PMACC_AUTO(particle, sourceFrame[localIdx]);
+
+        /* particle position, used for field-to-particle interpolation */
+        floatD_X pos = particle[position_];
+        const int particleCellIdx = particle[localCellIdx_];
+        /* multi-dim coordinate of the local cell inside the super cell */
+        DataSpace<TVec::dim> localCell(DataSpaceOperations<TVec::dim>::template map<TVec > (particleCellIdx));
+        /* interpolation of E-field on the particle position */
+        const fieldSolver::numericalCellType::traits::FieldPosition<FieldE> fieldPosE;
+        ValueType_E fieldE = Field2ParticleInterpolation()
+            (cachedE.shift(localCell).toCursor(), pos, fieldPosE());
+        /* interpolation of B-field on the particle position */
+        const fieldSolver::numericalCellType::traits::FieldPosition<FieldB> fieldPosB;
+        ValueType_B fieldB = Field2ParticleInterpolation()
+            (cachedB.shift(localCell).toCursor(), pos, fieldPosB());
+
+        const float3_X mom = particle[momentum_] / particle[weighting_];
+        const float_X mom2 = math::dot(mom, mom);
+        const float3_X mom_norm = mom * math::rsqrt(mom2);
+        const float_X mass = frame::getMass<FrameType>();
+
+        const float_X gamma = Gamma<>()(mom, mass);
+        const float3_X vel = mom / (gamma * mass); // low accuracy?
+
+        const float3_X lorentz = fieldE + math::cross(vel, fieldB);
+        const float_X lorentz2 = math::dot(lorentz, lorentz);
+        const float_X fieldE_long = math::dot(mom_norm, fieldE);
+
+        // effective magnetic strength
+        const float_X H_eff = math::sqrt(lorentz2 - fieldE_long*fieldE_long);
+
+        const float_X charge = math::abs(frame::getCharge<FrameType>());
+
+        const float_X c = SPEED_OF_LIGHT;
+        // Schwinger limit, unit: V/m
+        const float_X E_S = mass*mass * c*c*c / (charge * HBAR);
+        // quantum-nonlinearity parameter
+        const float_X chi = gamma * H_eff / E_S;
+
+        const float_X z = this->randomGen();
+
+        const float_X x = emission_prob_scaled(z, chi, gamma);
+
+        if(this->randomGen() < x)
+        {
+            const float_X photonMom_abs = z*z*z * mass*c * gamma;
+            if(photonMom_abs > SOFT_PHOTONS_CUTOFF_MOM)
+            {
+                this->photon_mom = mom_norm * photonMom_abs * particle[weighting_];
+                return 1;
+            }
+        }
+
+        return 0;
+    }
+
+    /** Functor implementation: setting photon and electron properties
+     *
+     * Called once for each single particle creation.
+     *
+     * \tparam Electron type of electron which creates the photon
+     * \tparam Photon type of photon that is created
+     */
+    template<typename Electron, typename Photon>
+    DINLINE void operator()(Electron& electron, Photon& photon) const
```

Comment:
This method could be called without calling `::numNewParticles` thus one might end up with a photon momentum of zero.

@psychocoderHPC @Heikman @ax3l Is it OK in PIConGPU to require these dependencies without checking them?